core/lib/object_store/src/s3.rs

```diff
@@ -0,0 +1,225 @@
+//! S3-based [`ObjectStore`] implementation.
+
+use std::fmt;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use aws_config::{meta::region::RegionProviderChain, BehaviorVersion, ConfigLoader, Region};
+use aws_runtime::env_config::file::{EnvConfigFileKind, EnvConfigFiles};
+use aws_sdk_s3::{error::SdkError, primitives::ByteStreamError, Client};
+use http::StatusCode;
+
+use crate::raw::{Bucket, ObjectStore, ObjectStoreError};
+
+/// [`ObjectStore`] implementation based on AWS S3.
+pub struct S3Store {
+    endpoint: String,
+    bucket_prefix: String,
+    client: Client,
+}
+
+impl fmt::Debug for S3Store {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("S3Store")
+            .field("bucket_prefix", &self.bucket_prefix)
+            // Skip `client` as its representation may contain sensitive info
```

Comment:
Nit: Is `endpoint` potentially sensitive as well? Otherwise, it could contain useful debug info.

Reply:
Added `endpoint` here. It isn't sensitive.